### ISSUE TITLE
trait & atoum : test it like we test class

### ISSUE THREAD

**Issue by Grummfy** [atoum/atoum#417]
It would be intresting to have a way to test a trait like we test class.

So the way to do it now, can be a class that use trait and we terst this class. But I'm pretty sure that we can find a way to make it simpler to test.

**Comment by Hywan** [comment]
A trait is not executable. What do you want to test?

**Comment by jubianchi** [comment]
@Hywan in itself a trait is not executable but we could run its method through a mock class using it.

**Comment by Grummfy** [comment]
You can create a mock on the fly with a trait usage inside.

so we will have a class like 

	class MyTrait extends \atoum\test
	{
		public function testSetterGetter()
		{
			$testedTrait = $this->getMockTrait('MyTrait');
			// or
			$testedTrait = new \mock\MyTrait();
			// or ...
			
			// do some assertion
			$testedTrait->setStuff('abc');
			$this->string($testedTrait->getStuff())->isEqualTo('abc');
		}
	}

**Comment by Hywan** [comment]
What about:
```php
trait T {
    public function getFoo() {
        return $this->_foo * 2;
    }
}
```
This is valid trait but a trait has not context by itself. The context is given by the implementer (the trait is the implementee). We basically can not test a trait because it is not testable by definition.

**Comment by Hywan** [comment]
To give a more complete answer:
```php
<?php

trait T {
    public function getFoo() {
        return $this->_foo * 2;
    }
}

class C {
    use T;

    protected $_foo = 21;
}

$c = new C();
var_dump($c->getFoo());
```

The context (`_foo`) is provided by `C` here. So you cannot test `T` but you can easily test `C`. The question is: How are you using traits?

**Comment by jubianchi** [comment]
I don't like when traits assume the implementer has some member variable. I prefer when they look like this:

```php
<?php

trait T {
    protected function compute($val) {
        return $val * 2;
    }
}

class C {
    use T;

    protected $_foo = 21;

    public function getFoo() {
        return $this->compute($this->_foo);
    }
}

$c = new C();
var_dump($c->getFoo());
```

The trait only provides utility to implementer, has not state and does not force the implementer to have a state in a given member. Its methods are hidden so not directly testable with atoum.

Here, when we run `C's test, we also test the trait but not directly.

**Comment by vonglasow** [comment]
@jubianchi +1

**Comment by Hywan** [comment]
@jubianchi What you like or dislike does not matter here (sorry :heart:), it is just what user can write and what atoum can do with it.

**Comment by mikaelrandy** [comment]
It's not because it's possible to write bad code that atoum should not be able to test it.

@Hywan your example is a piece of bad code (for me) and should not be write like this.
So, i think atoum should be able to test trait as it is classes.

As i tell to you on IRC, your exemple could be applied to on abstract class, but atoum allow to test abstract class.

**Comment by jubianchi** [comment]
@Hywan I just put an example of what I consider a good practice to illustrate how we can abstract implementation details in code and test fully test it.

@mikaelrandy atoum has always encouraged good practices through the tooling it provides. We won't likely introduce something allowing user to test bad designed code. Our goal is to help them get in a direction leading them to better design.

**Comment by mikaelrandy** [comment]
@jubianchi i agree with that ! Obviously !
Trait is not a bad designed code, on the contrary. And i think atoum should allow to test it to encourage it uses.

**Comment by jubianchi** [comment]
i did not say that trait arr wrong :) 

what I wanted to say is that they are/should be implementation details and so atoum should not be aware of them. they can still be tested through classes using them.

**Comment by mikaelrandy** [comment]
The point i want to show up is a trait could contain a public method, with some internal complexity who i want to test.
Consider this exemple :
```php
// Exposed in my bundle
trait ConsoleOutpout
{
  static public function output()
  {
    // Here is some transformation on output to be displayed on console
    printf($this->getOutput());
  }
}

// Use exemple
class ComputeSomeData extends SomeClass implements ConsoleOutput
{
  public function getOutput()
  {
    return $this->output;
  }
}
```

There is no implementation of my trait in my bundle, but i want to test my code.
Is there is any good reason to create a class only to test my trait ?
The code i want to test is my trait, not the class who implement my trait.

I only want to give an other point of view because i think the debate is deserved.
But there is just my point of view, and i will not debate for a week on this point ;)

**Comment by jubianchi** [comment]
@mikaelrandy fair point: the only valid use-case seems to be when someone wants to test a library which only provides common traits. 

what do you think if we provide tools for that particular use case in an extension dedicated to traits ? doing so, we won't add features covering "edge-cases" to the core.

**Comment by mikaelrandy** [comment]
hum ... extensions are they really used ?
I think if the functionnality not in the core, it will not be used. (i understand i could be wrong, but from my place, it's what i think ;))

But i let you decide what is the best for the atoum core.

**Comment by jubianchi** [comment]
let's just vote ;)

**Comment by mikaelrandy** [comment]
:+1:

**Comment by jubianchi** [comment]
So let's open the votes: should we introduce tools to test traits directly in atoum's core or in an extension ?

| Yes          | No             |
|--------------|----------------|
| :+1: (`:+1:`)| :-1: (`:-1:`)  |

End of the vote: ASAP, if this discussion does not get feedback for at least 1 week, I'll close it discussion and the corresponding issue.

For me it's a :-1::

> we won't add features covering "edge-cases" to the core.
>
> -- https://github.com/atoum/atoum/issues/417#issuecomment-125044707

**Comment by Hywan** [comment]
:-1:

**Comment by mageekguy** [comment]
:-1:

**Comment by mikaelrandy** [comment]
Yes : Core / No : extension ?

:+1:

**Comment by Hywan** [comment]
@mikaelrandy Exact.

**Comment by jubianchi** [comment]
@mikaelrandy I reformated my comment to be clear ;) thanks.

**Comment by mikaelrandy** [comment]
@mageekguy i agree, it a one-shot written too late in the night, i was too tired ;)

**Comment by Grummfy** [comment]
:-1:  => extension 

byt the way, don't forget this related PR: https://github.com/atoum/atoum/pull/432

**Comment by vonglasow** [comment]
:-1: => extension

**Comment by Oliboy50** [comment]
:+1: 

I agree with @mikaelrandy when he says that extension won't be used (besides I think it's a lot harder to promote a tool when its extra features are "hidden" or "somewhere else", and I think that Atoum is one of those tools which really have to be promoted).

But I don't know how much such a feature could impact the core, I just gave you my feelings about extensions.

The [visibility-extension](https://github.com/atoum/visibility-extension) for instance is an amazing stuff but because (I guess) it's an extension: this feature is not visible among all others in the README.

**Comment by Grummfy** [comment]
So I came back on this,
after reflexion, I came to the conclusion that it require four things :
* ability to generate a mock => simple (just made a simple change in the generator)
* assertion hasTrait() on asserter class => class_uses || ReflectionClass::getTraits
* assertion usingTrait() on asserter object (have not better name) => class_uses || ReflectionClass::getTraits
* nice error message when we try to use newTestedInstance => no idea on how to do it, yet

We can already create a class with unit test based on the trait itself.


So based on the current usage of traits, still believing it should be an extension?
